Refactored namespaces in builder

kobsy opened the pull request:
Before, namespace validation required traversing up the tree to find the ancestor with the namespace definition. When the ancestor is the document root and every element at every level is namespaced, this becomes problematic very quickly. So instead, namespace definitions are cached on children so that a child has only to look at its own definitions to validate a namespace. While it creates slightly more object overhead, the performance gains from getting rid of tree traversal are astronomical.

kobsy commented:
Ready for another look, @boblail

boblail commented on lib/openxml/builder.rb:
```diff
@@ -28,39 +28,24 @@ def to_s
       end
 
-      @ns = { :pending => ns.to_s }
+      @ns = OpenXml::Builder::Namespace.new(ns.to_s, nil)
```
Is a `OpenXml::Builder::Namespace` object valid without a `uri`?

kobsy replied:
It's not valid without a URI, but in order to have a valid URI, we'd have to use one of the workarounds we used previously to do validation (traversing all ancestors OR passing defined namespaces on to children). Since we're skipping validation, perhaps this doesn't need to be its own class anymore and could suffice to just be a symbol. I'll make that refactor.

boblail commented on lib/openxml/builder/element.rb:
```diff
@@ -21,17 +20,15 @@ def namespace=(ns)
       end
 
+      def namespaces=(ns)
+        @namespaces = ns if ns.is_a?(Array)
```
Why do you check if `ns.is_a?(Array)`? Is it ever not? Can we just assume that it is? Or should we coerce `ns` to an array with `Array(ns)` or allow it to cast itself to an array with `ns = ns.to_ary if ns.respond_to?(:to_ary)`?

boblail commented:
:+1: @kobsy. I like this refactor. I have two questions (not really **review-hold** questions, just questions-questions).

But, also, would you be willing to spec this with a few unit tests?

kobsy commented:
@boblail I refactored out the separate `OpenXml::Namespace` class, as it was primarily used only for validation, which we no longer do. I also added some unit tests around `OpenXml::Element`.
